parse/Parse.cpp

```diff
@@ -581,48 +359,181 @@ namespace parse {
     }
 
     namespace detail {
-        double_rule double_;
-
-        int_rule int_;
-
-        label_rule& label(const char* name) {
-            static const bool PARSING_LABELS_OPTIONAL = false;
-            static std::map<const char*, label_rule> rules;
-            std::map<const char*, label_rule>::iterator it = rules.find(name);
-            if (it == rules.end()) {
-                const lexer& lexer = lexer::instance();
-                label_rule& retval = rules[name];
-                if (PARSING_LABELS_OPTIONAL) {
-                    retval = -(lexer.name_token(name) >> '=');
-                } else {
-                    retval =  (lexer.name_token(name) >> '=');
-                }
-                retval.name(std::string(name) + " =");
-                return retval;
+    double_grammar::double_grammar(const parse::lexer& tok) :
```

Comment:
what's the distinction in naming between grammars and rules?

Reply:
A `rule` is a [boost spirit rule](http://www.boost.org/doc/libs/1_65_1/libs/spirit/doc/html/spirit/qi/reference/nonterminal/rule.html)

`rules` are a struct with a bunch of rules as members.  I did not name anything rules, nor did I fix the consistency of naming some things rules and others rule, when they are actually structs containing multiple rules.

A `grammar` is a [boost spirit grammar](http://www.boost.org/doc/libs/1_65_1/libs/spirit/doc/html/spirit/qi/reference/nonterminal/grammar.html), which is a set of rules with one rule designated as the start point.  The rule is designated in the constructor with `X::base_type(designated_start_rule_name)`.  The grammar otherwise, behaves like a simple rule.  Instead of typing `some_rules.special_rule_name`, type `some_grammar`.  Grammars are the built in `boost::spirit` way for composing more sophisticated rules.

The existing parser re-invented the grammar functionality in two places:
- All of the `parse::X()` functions were required to supply a `Rule` which was a structure containing at least one rule called `start`.  I replaced these with grammars, which can designate their own start point. 
- Some of the other rules were sets of with a single starting rule usually called `expr`, which was then referenced by the singleton that exposed those rules as a single rule.  I changed those to a grammar.

Additionally, in a few places there are structs that expose two or more rules as starting points.  I only fixed those that  were exposed as singletons in a way that prevents concurrent parsing.